src/modules/flow/flower-power/flower-power.c

```diff
@@ -0,0 +1,709 @@
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "flower-power-gen.h"
+#include "sol-flow-internal.h"
+
+#include <sol-http-client.h>
+#include <sol-json.h>
+#include <sol-util.h>
+
+#include <errno.h>
+#include <float.h>
+
+
+struct packet_type_flower_power_packet_data {
+    struct sol_drange fertilizer;
+    struct sol_drange light;
+    struct sol_drange temperature;
+    struct sol_drange water;
+    /* TODO: maybe convert it to int */
+    char *timestamp;
+    char *id;
+};
+
+static void
+packet_type_flower_power_packet_dispose(const struct sol_flow_packet_type *packet_type,
+                        void *mem)
+{
+    struct packet_type_flower_power_packet_data *packet_type_flower_power = mem;
+    free(packet_type_flower_power->id);
+}
+
+static int
+packet_type_flower_power_packet_init(const struct sol_flow_packet_type *packet_type,
+                     void *mem, const void *input)
+{
+    const struct packet_type_flower_power_packet_data *in = input;
+    struct packet_type_flower_power_packet_data *packet_type_flower_power = mem;
+
+    SOL_NULL_CHECK(in->id, -EINVAL);
+    packet_type_flower_power->id = strdup(in->id);
+    SOL_NULL_CHECK(packet_type_flower_power->id, -ENOMEM);
+
+    SOL_NULL_CHECK(in->timestamp, -EINVAL);
+    packet_type_flower_power->timestamp = strdup(in->timestamp);
+    SOL_NULL_CHECK(packet_type_flower_power->timestamp, -ENOMEM);
+
+    packet_type_flower_power->fertilizer = in->fertilizer;
+    packet_type_flower_power->light = in->light;
+    packet_type_flower_power->temperature = in->temperature;
+    packet_type_flower_power->water = in->water;
+
+    return 0;
+}
+
+#define PACKET_TYPE_FLOWER_POWER_PACKET_TYPE_API_VERSION (1)
+
+static const struct sol_flow_packet_type _PACKET_TYPE_FLOWER_POWER = {
```

Comment:
Do we really need a new packet type? Adding a new packet type for limited uses such as this is not a good idea IMO, considering that converter nodes will be needed to extract information from them, and that we lose things like console debugging or the inspector.